File: 63.unique-paths-ii/main.go
package main

func main() {
}

func uniquePathsWithObstacles(obstacleGrid [][]int) int {
	m := len(obstacleGrid)
	n := len(obstacleGrid[0])
	if obstacleGrid[0][0] == 1 || obstacleGrid[m-1][n-1] == 1 {
		return 0
	}
	grid := make([][]int, m)
	for i := range grid {
		grid[i] = make([]int, n)
	}
	grid[0][0] = 1
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if i > 0 && obstacleGrid[i-1][j] == 0 {
				grid[i][j] += grid[i-1][j]
			}
			if j > 0 && obstacleGrid[i][j-1] == 0 {
				grid[i][j] += grid[i][j-1]
			}
		}
	}
	return grid[m-1][n-1]
}

File: 63.unique-paths-ii/main.rs
impl Solution {
    pub fn unique_paths_with_obstacles(obstacle_grid: Vec<Vec<i32>>) -> i32 {
        let mut m = vec![vec![0; obstacle_grid[0].len()]; obstacle_grid.len()];
        if let Some(last_row) = m.last_mut() {
            if let Some(last_el) = last_row.last_mut() {
                *last_el = 1;
            }
        }

        for j in (0..obstacle_grid[0].len()).rev() {
            for i in (0..obstacle_grid.len()).rev() {
                if obstacle_grid[i][j] == 1 {
                    m[i][j] = 0;
                    continue;
                }
                if i < obstacle_grid.len() - 1 {
                    m[i][j] += m[i+1][j];
                }
                if j < obstacle_grid[0].len() - 1 {
                    m[i][j] += m[i][j+1];
                }
            }
        }
        m[0][0]
    }
}
